ticdc: version 8.3.0 of TiCDC is not compatible with version 8.1.0 of TiKV

### What did you do?

**Description**:  
While performing integration testing with TiCDC's master branch (version 8.3.0), I encountered an issue where the `batch_add_table` test fails. The TiCDC node crashes because the puller module panics during DDL execution.

**Steps to Reproduce**:
1. Run `make prepare_test_binaries community=true` to download TiDB, TiKV, PD, and other components. This pulls version 8.1.0 of the components (I'll explain why below).
2. Execute `make integration_test_build`.
3. Execute `make integration_test` to start the integration test.

**Observed Behavior**:  
During the `batch_add_table` integration test, TiCDC's puller module panics when processing DDL, causing the TiCDC node to crash.

**Cause Analysis**:  
The reason TiCDC 8.3.0 pulled version 8.1.0 of TiDB, TiKV, PD, etc., is that when running `make prepare_test_binaries` with the `community=true` option, if the version is not manually specified, it defaults to downloading version 8.1.0 of all TiDB components. (related to #11508)

After manually replacing the TiDB, TiKV, PD components with the latest versions, the `batch_add_table` integration test passed successfully. This indicates that TiCDC 8.3.0 currently does not maintain backward compatibility with TiKV 8.1.0.


### What did you expect to see?

I expected the batch_add_table integration test to pass without any issues when using TiCDC 8.3.0 with TiDB, TiKV, and PD components (even if they are from version 8.1.0). The system should handle version differences gracefully, or at the very least, the prepare_test_binaries step should automatically pull compatible versions of all components.

### What did you see instead?


![img_v3_02e0_1c59243e-9039-4edf-bbf7-7592e91e009g](https://github.com/user-attachments/assets/51433ffd-72b6-440e-bbf7-816a9f80e5da)


```
panic: runtime error: index out of range [1] with length 1

goroutine 809 [running]:
[github.com/pingcap/tiflow/cdc/puller.(*ddlJobPullerImpl).handleJob](http://github.com/pingcap/tiflow/cdc/puller.(*ddlJobPullerImpl).handleJob)(0xc0040b7980, 0xc001756820)
        /root/tiflow/cdc/puller/ddl_puller.go:410 +0x4c92
[github.com/pingcap/tiflow/cdc/puller.(*ddlJobPullerImpl).handleRawKVEntry](http://github.com/pingcap/tiflow/cdc/puller.(*ddlJobPullerImpl).handleRawKVEntry)(0xc0040b7980, {0xbde3480, 0xc0031b5db0}, 0xc00123b1f0)
        /root/tiflow/cdc/puller/ddl_puller.go:206 +0x35c
[github.com/pingcap/tiflow/cdc/puller.(*ddlJobPullerImpl).Run.func3()](http://github.com/pingcap/tiflow/cdc/puller.(*ddlJobPullerImpl).Run.func3())
        /root/tiflow/cdc/puller/ddl_puller.go:153 +0x218
[golang.org/x/sync/errgroup.(*Group).Go.func1()](http://golang.org/x/sync/errgroup.(*Group).Go.func1())
        /root/go/pkg/mod/[golang.org/x/sync@v0.7.0/errgroup/errgroup.go:78](http://golang.org/x/sync@v0.7.0/errgroup/errgroup.go:78) +0x92
created by [golang.org/x/sync/errgroup.(*Group).Go](http://golang.org/x/sync/errgroup.(*Group).Go) in goroutine 805
        /root/go/pkg/mod/[golang.org/x/sync@v0.7.0/errgroup/errgroup.go:75](http://golang.org/x/sync@v0.7.0/errgroup/errgroup.go:75) +0x125
~
```

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
8.1.0
```

Upstream TiKV version (execute `tikv-server --version`):

```console
8.1.0
```

TiCDC version (execute `cdc version`):

```console
8.3.0
```

--- a/cdc/puller/ddl_puller.go
+++ b/cdc/puller/ddl_puller.go
@@ -396,11 +396,12 @@ func (p *ddlJobPullerImpl) handleJob(job *timodel.Job) (skip bool, err error) {
 		// we only use multiTableInfos and Querys when we generate job event
 		// So if some table should be discard, we just need to delete the info from multiTableInfos and Querys
 		if strings.Count(job.Query, ";") != len(job.BinlogInfo.MultipleTableInfos) {
-			log.Panic("the number of queries in `Job.Query` is not equal to "+
+			log.Error("the number of queries in `Job.Query` is not equal to "+
 				"the number of `TableInfo` in `Job.BinlogInfo.MultipleTableInfos`",
 				zap.String("Job.Query", job.Query),
 				zap.Any("Job.BinlogInfo.MultipleTableInfos", job.BinlogInfo.MultipleTableInfos),
 				zap.Error(cerror.ErrTiDBUnexpectedJobMeta.GenWithStackByArgs()))
+			return false, cerror.ErrTiDBUnexpectedJobMeta.GenWithStackByArgs()
 		}
 
 		var newMultiTableInfos []*timodel.TableInfo

--- a/cdc/puller/ddl_puller.go
+++ b/cdc/puller/ddl_puller.go
@@ -397,7 +397,7 @@ func (p *ddlJobPullerImpl) handleJob(job *timodel.Job) (skip bool, err error) {
 		// So if some table should be discard, we just need to delete the info from multiTableInfos and Querys
 		if strings.Count(job.Query, ";") != len(job.BinlogInfo.MultipleTableInfos) {
 			log.Panic("the number of queries in `Job.Query` is not equal to "+
-				"the number of `TableInfo` in `Job.BinlogInfo.MultipleTableInfos`, ",
+				"the number of `TableInfo` in `Job.BinlogInfo.MultipleTableInfos`",
 				zap.String("Job.Query", job.Query),
 				zap.Any("Job.BinlogInfo.MultipleTableInfos", job.BinlogInfo.MultipleTableInfos),
 				zap.Error(cerror.ErrTiDBUnexpectedJobMeta.GenWithStackByArgs()))

--- a/cdc/puller/ddl_puller.go
+++ b/cdc/puller/ddl_puller.go
@@ -395,6 +395,14 @@ func (p *ddlJobPullerImpl) handleJob(job *timodel.Job) (skip bool, err error) {
 	case timodel.ActionCreateTables:
 		// we only use multiTableInfos and Querys when we generate job event
 		// So if some table should be discard, we just need to delete the info from multiTableInfos and Querys
+		if strings.Count(job.Query, ";") != len(job.BinlogInfo.MultipleTableInfos) {
+			log.Panic("the number of queries in `Job.Query` is not equal to "+
+				"the number of `TableInfo` in `Job.BinlogInfo.MultipleTableInfos`, ",
+				zap.String("Job.Query", job.Query),
+				zap.Any("Job.BinlogInfo.MultipleTableInfos", job.BinlogInfo.MultipleTableInfos),
+				zap.Error(cerror.ErrTiDBUnexpectedJobMeta.GenWithStackByArgs()))
+		}
+
 		var newMultiTableInfos []*timodel.TableInfo
 		var newQuerys []string
 
@@ -407,13 +415,13 @@ func (p *ddlJobPullerImpl) handleJob(job *timodel.Job) (skip bool, err error) {
 				continue
 			}
 			newMultiTableInfos = append(newMultiTableInfos, multiTableInfos[index])
-			newQuerys = append(newQuerys, querys[index])
+			newQuerys = append(newQuerys, querys[index]+";")
 		}
 
 		skip = len(newMultiTableInfos) == 0
 
 		job.BinlogInfo.MultipleTableInfos = newMultiTableInfos
-		job.Query = strings.Join(newQuerys, ";")
+		job.Query = strings.Join(newQuerys, "")
 	case timodel.ActionRenameTable:
 		oldTable, ok := snap.PhysicalTableByID(job.TableID)
 		if !ok {

--- a/errors.toml
+++ b/errors.toml
@@ -1011,6 +1011,11 @@ error = '''
 fail to create changefeed because target-ts %d is earlier than start-ts %d
 '''
 
+["CDC:ErrTiDBUnexpectedJobMeta"]
+error = '''
+unexpected `job_meta` from tidb
+'''
+
 ["CDC:ErrTiKVEventFeed"]
 error = '''
 tikv event feed failed

--- a/pkg/errors/cdc_errors.go
+++ b/pkg/errors/cdc_errors.go
@@ -145,6 +145,12 @@ var (
 		errors.RFCCodeText("CDC:ErrCodeNilFunction"),
 	)
 
+	// Errors caused by unexpected behavior from external systems
+	ErrTiDBUnexpectedJobMeta = errors.Normalize(
+		"unexpected `job_meta` from tidb",
+		errors.RFCCodeText("CDC:ErrTiDBUnexpectedJobMeta"),
+	)
+
 	// ErrVersionIncompatible is an error for running CDC on an incompatible Cluster.
 	ErrVersionIncompatible = errors.Normalize(
 		"version is incompatible: %s",

